Resource Bundle Look-ups in Modular Java Applications (#31)
gunnarmorling/morling.dev: content/blog/resource-bundle-lookups-in-modular-java-applications.asciidoc

}
----

The name of bundle provider interfaces must follow the pattern `<package of baseName> + ".spi." + <simple name of baseName> + "Provider"`.
As the base name is `dev.morling.links.core.LinksMessages` in our case, the provider interface name must be `dev.morling.links.core.spi.LinksMessagesProvider`.
This can be sort of a stumbling stone, as an innocent typo in the package or type name will cause your bundle not to be found,
without good means of analyzing the situation, other than double and triple checking that all names are correct.

Next, we need to declare the usage of this provider interface in the consuming module.
Assuming the afore-mentioned error handler class is located in the _core_ module,
the module descriptor of the same looks like so:

[source,java,linenums=true]
----
module dev.morling.links.core {
  exports dev.morling.links.core;
  exports dev.morling.links.core.spi; // <1>
  uses dev.morling.links.core.spi.LinksMessagesProvider; // <2>
}
----
<1> Export the package of the resource bundle provider interface so that implementations can be created in other modules
<2> Declare the usage of the `LinksMessagesProvider` service

> Using the resource bundle in the error handler class is rather un-exciting;

Author: ```suggestion
Using the resource bundle in the error handler class is rather unexciting;
```

> Using the resource bundle in the error handler class is rather un-exciting;

Author: Thx, changed.

Once the look-up code in `ResourceBundle` has found a provider which returns a bundle for a requested name and locale,
it will not query any other bundle providers.
In our case though, we need to be able to obtain messages from any of the bundles contributed by the different modules:
messages related to green keeping must be obtained from the bundle of the `dev.morling.links.greenkeeping` module,
tournament messages from `dev.morling.links.tournament`, and so on.

The idea to address this concern is the following:

* Prefix each message key with a module specific string, resulting in keys like `tournament.fullybooked`, `greenkeeping.greenclosed`, etc.
* When requesting the bundle for a given key in the error handler class, obtain the key's prefix and pass it to bundle providers
* Let bundle providers react only to their specific message prefix

This is where things become a little bit fiddly:
there isn't a really good way for passing such contextual information from bundle consumers to providers.
Our loop hole here will be to squeeze that information into the the requested https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Locale.html[`Locale`] instance.

> Besides the well-known language and country attributes, `Locale` can also carry _variant_ data and even application specific https://docs.oracle.com/javase/tutorial/i18n/locale/extensions.html[_extensions_].

Author: Yes, that'd be the idea. If it worked :) I'll clarify it.

The latter would actually be pretty much ideal for our purposes.
But unfortunately, extensions aren't evaluated by the look-up routine in `ResourceBundle`.
So instead we'll go with propagating the key namespace information via the locale's variant.
First, let's revisit the code in the `ErrorHandler` class:

[source,java,linenums=true]
----
public class ErrorHandler {

  public String getErrorMessage(String key, UserContext context) {

>       String component = key.split("\\.")[0]; // <1>

Reviewer: nit pick -> why is the var named 'component'. if I didn't miss it the text never talks about components but bundles, modules and key prefixes?

Author: Good catch, I was/am using the terms component and modules somewhat interchangeable. I'll adjust it.

>   public String getErrorMessage(String key, UserContext context) {
>       String component = key.split("\\.")[0]; // <1>
>       Locale locale = new Locale( // <2>
>           context.getLocale().getLanguage(),
>           context.getLocale().getCountry(),
>           component
>       );

Author: ```suggestion
  public String getErrorMessage(String key, UserContext context) {
      String prefix = key.split("\\.")[0]; // <1>
      Locale locale = new Locale( // <2>
          context.getLocale().getLanguage(),
          context.getLocale().getCountry(),
          prefix
      );
```

[source,xml,linenums=true]
----
...
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-compiler-plugin</artifactId>
  <executions>
    <execution> // <1>
      <id>compile-java-9</id>
      <phase>compile</phase>
      <goals>
        <goal>compile</goal>
      </goals>
      <configuration>
        <release>9</release> // <2>
        <compileSourceRoots>
          <compileSourceRoot>
            ${project.basedir}/src/main/java-9 // <3>
          </compileSourceRoot>
        </compileSourceRoots>
      <multiReleaseOutput>true</multiReleaseOutput> // <4>
    </configuration>
    </execution>
  </executions>
</plugin>
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-jar-plugin</artifactId>
  <configuration>
    <archive>
      <manifestEntries>
        <Multi-Release>true</Multi-Release> // <5>
      </manifestEntries>
    </archive>
  </configuration>
</plugin>
...
----
<1> Set up another execution of the https://maven.apache.org/plugins/maven-compiler-plugin/[Maven compiler plug-in] for the Java 9 specific sources,
<2> using Java 9 bytecode level,
<3> picking up the sources from _src/main/java-9_,

> <4> and organizing the compilation output in the multi-relese structure under _META-INF/versions/..._

Reviewer: multi-rele*a*se

> <4> and organizing the compilation output in the multi-relese structure under _META-INF/versions/..._

Author: ```suggestion
<4> and organizing the compilation output in the multi-release structure under _META-INF/versions/..._
```

* _No knowledge about the specific modules in the central component:_ ✅  Thanks to the different providers (or the custom `Control`, respectively), there is no need for registering the specific bundles with the error handler in the _core_ module; further modules could be added to the Links application and the error handler would be able to obtain messages from the resource bundles contributed by them

With a little bit of extra effort, it also was possible to design the code in the _core_ module in a way that the application can be used with different Java versions and configurations:
on the module path with Java 9+, on the classpath with Java 9+, on the classpath with Java 8.

If you'd like to explore the complete code by yourself, you can find it in the https://github.com/gunnarmorling/modular-resource-bundles[modular-resource-bundles] GitHub repository.
To learn more about resource bundle retrieval in named modules,

> please refer to the extensive documentation of https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ResourceBundle.html#resource-bundle-modules[`ResourceBundle`] and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/spi/ResourceBundleProvider.html[`ResourceBundleProvider`].

Author: ```suggestion
please refer to the extensive documentation of https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ResourceBundle.html#resource-bundle-modules[`ResourceBundle`] and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/spi/ResourceBundleProvider.html[`ResourceBundleProvider`].

_Many thanks to https://twitter.com/hpgrahsl[Hans-Peter Grahsl] for providing feedback while writing this post!_
```
